File: src/main/rs/emulate/legacy/IndexCodec.java
package rs.emulate.legacy;

import rs.emulate.shared.util.DataBuffer;

import com.google.common.base.Preconditions;

/**
 * Contains methods for encoding and decoding {@link Index} objects.
 * 
 * @author Major
 * @author Graham
 */
final class IndexCodec {

	/**
	 * Decodes the {@link DataBuffer} into an index.
	 * 
	 * @param buffer The buffer.
	 * @return The index.
	 * @throws IllegalArgumentException If the buffer length is invalid.
	 */
	public static Index decode(DataBuffer buffer) {
		Preconditions.checkArgument(buffer.remaining() == Index.BYTES, "Incorrect buffer length.");

		int size = buffer.getUnsignedTriByte();
		int block = buffer.getUnsignedTriByte();

		return new Index(size, block);
	}

	/**
	 * Encodes the index into a {@link DataBuffer}.
	 * 
	 * @param index The index.
	 * @return The byte buffer.
	 */
	public static DataBuffer encode(Index index) {
		return DataBuffer.allocate(6).putTriByte(index.getSize()).putTriByte(index.getBlock()).flip();
	}

	/**
	 * Default private constructor to prevent instantiation.
	 */
	private IndexCodec() {

	}

}

File: legacy/src/main/kotlin/rs/emulate/legacy/IndexCodec.kt
package rs.emulate.legacy

import io.netty.buffer.ByteBuf
import io.netty.buffer.Unpooled
import rs.emulate.util.readUnsignedTriByte
import rs.emulate.util.writeTriByte
import java.nio.ByteBuffer

/**
 * Contains methods for encoding and decoding [Index] objects.
 */
internal object IndexCodec {

    /**
     * Decodes the [ByteBuffer] into an index.
     * @throws IllegalArgumentException If the buffer length is invalid.
     */
    fun decode(buffer: ByteBuf): Index {
        require(buffer.readableBytes() == Index.BYTES) { "Incorrect buffer length." }

        val size = buffer.readUnsignedTriByte()
        val block = buffer.readUnsignedTriByte()

        return Index(size, block)
    }

    /**
     * Encodes the index into a [ByteBuffer].
     */
    fun encode(index: Index): ByteBuf {
        return Unpooled.buffer(6).apply {
            writeTriByte(index.size)
            writeTriByte(index.block)
        }
    }

}
